### PR TITLE
Update EIP-6963: Add `providerId` into the design

### PR DESCRIPTION
**Pull request by jxom**
The [wagmi](https://github.com/wagmi-dev) core team proposes to add a providerId into the design of EIP-6963, formally known as `walletId`.

Decided not to use an rdns (Reverse DNS) for the said deterministic identifier to keep the objective clear, focused and simple, without having to deal with the complexities of the rationale of a rDNS in this spec and potential side-effects of a rDNS that may impact the determinism of the identifier in the future.

Update: [changed `walletId` to `providerId`](https://github.com/ethereum/EIPs/pull/7167/commits/552cb791ca9880991b7f06928fe2976bf0bd1d8e) (ref: https://ethereum-magicians.org/t/eip-6963-multi-injected-provider-discovery/14076/23).

### CODE REVIEW

File: EIPS/eip-6963.md [hunk 1]
```diff
@@ -39,27 +39,72 @@ Wallet: A user agent that manages keys and facilitates transactions with Ethereu
 
 Decentralized Application (DApp): A web page that relies upon one or many Web3 platform APIs which are exposed to the web page via the Wallet.
 
-Provider Discovery Library: A library or piece of software that assists a DApp to interact with the Wallet. 
+Provider Discovery Library: A library or piece of software that assists a DApp to interact with the Wallet (e.g. wagmi, Web3Modal, RainbowKit, Web3-Onboard, ConnectKit, etc).
```
Comment (jxom) [inline EIPS/eip-6963.md:42]: Moved from L172 to here to make it clear on examples of Provider Discovery Libraries. Readers might need an example to properly piece it together before moving on.

File: EIPS/eip-6963.md [hunk 2]
```diff
@@ -39,27 +39,72 @@ Wallet: A user agent that manages keys and facilitates transactions with Ethereu
...
 
 ### Provider Info
 
-Each wallet provider will be announced with the following interface `EIP6963ProviderInfo`. The values in the `EIP6963ProviderInfo` MUST be used as follows:
+Each Wallet Provider will be announced with the following interface `EIP6963ProviderInfo`. The values in the `EIP6963ProviderInfo` MUST be used as follows:
 
-- **`uuid`** - a globally unique of the wallet provider that MUST be (UUIDv4[^rfc4122] compliant) to uniquely distinguish different [EIP-1193](./eip-1193.md) provider sessions. The cryptographic uniqueness provided by UUIDv4[^rfc4122] guarantees that two independent `EIP6963ProviderInfo` objects can be separately identified.
```
Comment (jxom) [inline EIPS/eip-6963.md:48]: Just moved this line down to "Universally Unique Identifiers (UUIDs)" underneath (to keep consistent with the "Wallet IDs", "Wallet Names" and "Images/Icons" format)

**Review by Arschieqwe — APPROVED** [review]

**Comment by lochie** [comment]
Lochie here, lead dev on [ConnectKit](https://github.com/family/connectkit).

`providerId`  **should definitely** be included in this proposal. Without it we would be left to detect/target providers based on the `name` attribute which isn't really a standard enough identifier to be 100% sure that it's what we're intending to find.

`providerId` would also be useful for if providers change their branding name (i.e `Gnosis Safe` rebranded to `Safe`), so the identifier could remain the same until project/library migrations have caught up, but the branding can be more stably swapped out if necessary.

tl;dr: `providerId` yes pls 🙏

**Review by kdenhartog — CHANGES_REQUESTED** [review]
As it stands we've implemented a ton of non-normative text to this without defining "HOW" it should work. Without including this detail it will lead to a centralization of behavior around the most supported libraries being copied just as we saw happen with new wallets copying metamask's behavior. This spec is doing a really good thing to move us away from using early design choices as defacto standards and if we are going to support a providerId we should be defining normative requirements to achieve this.

File: EIPS/eip-6963.md [hunk 3]
```diff
@@ -39,27 +39,72 @@ Wallet: A user agent that manages keys and facilitates transactions with Ethereu
 
 Decentralized Application (DApp): A web page that relies upon one or many Web3 platform APIs which are exposed to the web page via the Wallet.
 
-Provider Discovery Library: A library or piece of software that assists a DApp to interact with the Wallet. 
+Provider Discovery Library: A library or piece of software that assists a DApp to interact with the Wallet (e.g. Web3Modal, wagmi, RainbowKit, Web3-Onboard, ConnectKit, etc).
```
[the remark this answers is not in the document]
Comment (jxom) [reply EIPS/eip-6963.md:42]: Was mentioned later on in the spec, but can remove it.

File: EIPS/eip-6963.md [hunk 4]
```diff
@@ -39,27 +39,72 @@ Wallet: A user agent that manages keys and facilitates transactions with Ethereu
...
 
-Each wallet provider will be announced with the following interface `EIP6963ProviderInfo`. The values in the `EIP6963ProviderInfo` MUST be used as follows:
+Each Wallet Provider will be announced with the following interface `EIP6963ProviderInfo`. The values in the `EIP6963ProviderInfo` MUST be used as follows:
 
-- **`uuid`** - a globally unique of the wallet provider that MUST be (UUIDv4[^rfc4122] compliant) to uniquely distinguish different [EIP-1193](./eip-1193.md) provider sessions. The cryptographic uniqueness provided by UUIDv4[^rfc4122] guarantees that two independent `EIP6963ProviderInfo` objects can be separately identified.
-- **`name`** - A human-readable local alias of the wallet provider to be displayed to the user on the DApp. (e.g. `DopeWalletExtension` or `AwesomeWallet`)
-- **`icon`** - A URI[^rfc3986] pointing to an image. Icon images MUST be square with 96x96px minimum resolution
+- **`uuid`** - a universally unique identifier of the Wallet Provider that MUST be UUIDv4[^rfc4122] compliant and thus MUST BE nondeterministic between sessions.
+- **`providerId`** - a unique identifier for the Wallet Provider that MUST be deterministic between sessions.
```
[the remark this answers is not in the document]
Comment (jxom) [reply EIPS/eip-6963.md:49]: It's determinsitic between sessions if the identifier is identical between sessions. Should we define what "determinism" and "session" means Definitions?

File: EIPS/eip-6963.md [hunk 5]
```diff
@@ -169,10 +214,31 @@ To follow the Javascript event name conventions, the names are written in presen
...
+
+Furthermore, some use-cases of Provider IDs for Provider Discovery Libraries include:
+
+- **Rehydrating Wallet Providers**: If the Provider Discovery Library has persisted the Wallet Provider in a previous session via browser local storage or a database, they can rely on the Provider ID to accurately rehydrate the Wallet Provider. Same applies for if a Provider Discovery Library emits the `"eip6963:requestProvider"` event again in the same session.
+- **Group Wallets**: If the Provider Discovery Library has a list of Wallet Providers they wish to group together (e.g. a "Recent", "Favourites", etc section) they can rely on the Provider ID to accurately achieve this.
+- **`wallet_` RPC Method Differences**: If the Wallet Provider implements optional `wallet_` RPC methods, the Provider Discovery Library can rely on the Provider ID to accurately determine how to facilitate the `wallet_` RPC method. For example, a Wallet may not support chain switching, and thus may not support `wallet_switchEthereumChain`, this means the Provider Discovery Library could not show a "Switch Network" button for that Wallet Provider.
```
[the remark this answers is not in the document]
Comment (jxom) [reply EIPS/eip-6963.md:241]: I ideally think that these RPC methods should be defined as a feature set on the Wallet (a la. CAIP-25), however, this is obviously not in scope for this proposal. I just added this as an example of why a Provider Discovery Library needs a deterministic ID to identify the Wallets that may not implement these methods.

File: EIPS/eip-6963.md [hunk 6]
```diff
@@ -169,10 +214,31 @@ To follow the Javascript event name conventions, the names are written in presen
...
+Previously, we implicitly achieved Provider IDs by extending `window.ethereum` to include a non-standard flag (ie. `isMetaMask`, `isBrave`, `isTrustWallet`, etc.) to identify Wallet Providers. This was problematic because it:
+
+1. is a non-standard
+2. introduces contradicting/"impossible" states (e.g. `isMetaMask` and `isOpera` being both `true`) that lead to challenges when identifying the correct wallet.
+
+Having a stringified Provider ID (`providerId`) instead of a boolean flag allows us to accurately identify Wallet Providers between sessions and avoid the aforementioned challenges.
+
+Furthermore, some use-cases of Provider IDs for Provider Discovery Libraries include:
+
+- **Rehydrating Wallet Providers**: If the Provider Discovery Library has persisted the Wallet Provider in a previous session via browser local storage or a database, they can rely on the Provider ID to accurately rehydrate the Wallet Provider. Same applies for if a Provider Discovery Library emits the `"eip6963:requestProvider"` event again in the same session.
```
[the remark this answers is not in the document]
Comment (jxom) [reply EIPS/eip-6963.md:239]: This is an example of why Provider Discovery Libraries need a deterministic identifier. It doesn't necessarily need to be standardized, no.

File: EIPS/eip-6963.md [hunk 7]
```diff
@@ -39,27 +39,72 @@ Wallet: A user agent that manages keys and facilitates transactions with Ethereu
...
   uuid: string;
+  providerId: string;
   name: string;
   icon: string;
 }
 ```
 
+#### Universally Unique Identifiers (UUIDs)
+
+The UUID (`uuid`) is a **nondeterministic unique identifier** used to uniquely distinguish different [EIP-1193](./eip-1193.md) provider sessions. The cryptographic uniqueness provided by UUIDv4[^rfc4122] guarantees that two independent `EIP6963ProviderInfo` objects can be separately identified.
```
[the remark this answers is not in the document]
Comment (jxom) [reply EIPS/eip-6963.md:67]: Can add to Definitions.

**Comment by jxom** [comment]
I appreciate you taking the time out to review this and I can address your comments, as well as elaborating more on the HOW a determinsitic identifier should work. 

We, library authors, need a deterministic identifier like `providerId` in order to implement a robust implementation of the proposal (as justified in the Rationale). Without this identifier, we will be relying on the Provider Name for determinism, which may not be deterministic by nature (it's a display name that could be subject to change and could be localized). Or even worse, the flags on `window.ethereum`.

Without a deterministic ID, this will lead to centralization of library-defined deterministic IDs → name as libraries will want to prevent their consumers (DApps) from being impacted by the unpredicted breaking change whenever a Wallet modifies/localizes their name. Wallet brand/name changes happen often (we have seen [three](https://github.com/wagmi-dev/references/commit/bbbd11b82a8ca2b7347bfbc9972999f152e7a777) [instances](https://github.com/wagmi-dev/references/commit/0778abc80294058865771d1c6ca9d9ec47b74050) of [changes](https://github.com/wagmi-dev/references/commit/0966bf7173dfa644a3bcadb5fbd1ef978501c11a) in the past few months, and more beyond).

Hope you understand where I am coming from. If this purely just needs more justification on the "HOW", I am willing to amend.

File: EIPS/eip-6963.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (kdenhartog) [reply EIPS/eip-6963.md:42]: Yeah I noticed that as well, I think I left that in there accidentally and only just noticed it now when reviewing.

File: EIPS/eip-6963.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (kdenhartog) [reply EIPS/eip-6963.md:49]: Probably if we want to distinguish between what they mean. Having clear definitions for things like this can make it easier for the readers and implementors to understand why each one is necessary.

[the remark this answers is not in the document]
Comment (kdenhartog) [reply EIPS/eip-6963.md:49]: That was just my understanding of what the consensus was in the Telegram group. I wasn't aware we were now requiring it.

File: EIPS/eip-6963.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (lochie) [reply EIPS/eip-6963.md:42]: +1 spec should be agnostic

File: EIPS/eip-6963.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (jxom) [reply EIPS/eip-6963.md:49]: Yep agree. We could also probably rename "session" to "page instance" or something more specific to the browser.

**Comment by DanielSinclair** [comment]
RainbowKit team here 🫡

`providerId` is an important addition for us:
- We have found that many injected wallet providers prefer a different display name than their connector `id`. Usually this looks like an `id` of `best` for `Best Wallet`. This is even more complicated by the flag standardization today for wallets like `Core Wallet` which relies on `isAvalanche`, or 1inch Wallet or Trust Wallet that have different provider flags between their supported platforms. In RainbowKit, we do a little extra work for a better UX by tying these together under the wallet's brand, and handling different behavior on different platforms. A stable `projectId` would help us achieve this.
- We have also coordinated a number of wallet rebrands, and decoupling the display name from the `providerId` at the very least would be very helpful for this scenario. Allowing wallets to register multiple `providerId`s to assist a supply chain transition would be helpful as well. This is akin to how a rebranded iOS app can continue to register a legacy deep link while introducing a new display name and deep link.

Thank you @jxom for spotting and recommending this!

**Comment by glitch-txs** [comment]
Adding to this on how I see it:
- Having a non-deterministic `uuid` is great for filtering wallets on a page load, it's useful to make sure we are not filtering wallets by mistake because of impersonation.
- Having a deterministic `providerId` is great for keeping track of and saving information of previews sessions on client storage.
- And the wallet's `name` is good for displaying it on the UI

**Comment by kdenhartog** [comment]
@jxom and I hopped on a call yesterday about this PR. In short, I can see how this would be helpful for libraries but we also concluded that this wasn't going to behave deterministically in the way we'd want to unless we ran down the rabbithole of trying to define all of the undefined behavior. Instead, the conclusion was that we'd make this an additional optional parameter while still defining the format of the identifier if it's supplied so that it can act as an intentional extension point for DApp/library authors to implement their own behavior without needing to standardize on the behavior of that for libraries and DApps. This means this property in this spec would act very similar to how OAuth2 goes about defining the tokens that get used within the protocol. Most people would expect that it's a JWT, but in the spec it's intentional left vaguely defined so that each implementation can utilize whichever specific token type they choose. See the last paragraph of this [section](https://datatracker.ietf.org/doc/html/rfc6749#section-1.4) to understand what I mean here.

So, the next steps are on me to update this PR and clean up the spec language to align with this and once that's done we'll poll the TG group to see if we've got consensus to go in this direction.

**Comment by tmm** [comment]
@kdenhartog

> we also concluded that this wasn't going to behave deterministically in the way we'd want to unless we ran down the rabbithole of trying to define all of the undefined behavior.

@jxom, @lochie, and @DanielSinclair both presented compelling points for making `providerId` required. Seems worth going down the rabbithole to define the "undefined behavior" you are talking about.

The only case I can think of is two wallets setting `providerId` to the same value (either by accident or on purpose). Seems like wallets can mostly avoid this in the same way chain builders avoid choosing the same chain ID. (Could also keep a GitHub registry of Wallet Providers.)

Would love to hear about other cases I might be missing!

> we'll poll the TG group to see if we've got consensus to go in this direction.

Would be great if this process could be public and open, rather than private and centralized (not everyone uses Telegram). I'm concerned that library authors' feedback isn't being included in the process and this API getting designed in a vacuum.

—

Overall, at the very least, offering concrete reasons for why this change should not move forward is important.

The library authors commenting and supporting this change build core ecosystem software that is used by thousands of developers and dApps, like ENS, Optimism, WalletConnect, Stripe, Shopify, PancakeSwap, Sushi, LooksRare, and many more. If EIPs won't consider their feedback, who is this process for?

**Comment by lochie** [comment]
agreeing with @tmm here. this should be aiming for future-proofing and conflict avoiding.

**Comment by kdenhartog** [comment]
Do you have specific spec text that you'd like to suggest for this PR or https://github.com/ethereum/EIPs/pull/7256?

I've put forth an alternative proposal which @jxom has approved so I'd prefer to get more concrete about the changes you're hoping to see added to the spec if you believe that what I've proposed is not sufficient.

> The only case I can think of is two wallets setting providerId to the same value (either by accident or on purpose). Seems like wallets can mostly avoid this in the same way chain builders avoid choosing the same chain ID. (Could also keep a GitHub registry of Wallet Providers.)

My argument for this from the beginning has been that we should avoid a centralized registry by any means necessary. We need to consider the fact that these core libraries would behave as psuedo-central registries as well because of how commonly used they are by many DApps. The reason we should avoid this is because conflict is inevitable with any registry and therefore requires some mechanism of governance to be maintained. The EIP process is not conducive to managing a registry so we'd then need to create our own governance process to resolve disagreements such as namespace collisions, modifications to entries, or removal of entries for some reason. As someone who's maintained registries at W3C and participated in registering various entries at IANA for content types and the JOSE registry I want to avoid them at all costs. They always introduce friction and quite commonly re-enforce the "private and centralized" communication channels that you're already concerned that we're faced with.

 > Would be great if this process could be public and open, rather than private and centralized (not everyone uses Telegram). I'm concerned that library authors' feedback isn't being included in the process and this API getting designed in a vacuum.
 
 On this topic, I agree this is sorely needed within the entire Ethereum ecosystem. Unfortunately the EIP process sets out that the discussions are supposed to happen at https://ethereum-magicians.org, but for the most part none of the discussion has actually been happening there. The majority of people participating have been conducting the discussions either via the Telegram channel or the AllWalletDevs discord channel. Since we don't formally have any sort of membership or participation within the EIP process this Telegram channel has acted as a psuedo working group channel for us to move things forward for the majority of the relevant parties. Personally, I'm not a fan of either Telegram or Discord but if that's where the discussions are happening then I'll show up. I vastly prefer weekly/bi-weekly calls and github issues to move things forward. With that said, I have no intention of ignoring the feedback you're providing and am noting that there's reservations from DApp library developers which is important to consider. From the beginning I noticed that this spec was largely being driven by Wallet Devs so it's good that we're seeing participation from the DApp side as well. I'd love to see more actually DApp developers participating in these discussions too, but we'll take what we can get.
 
> Overall, at the very least, offering concrete reasons for why this change should not move forward is important.
> 
> The library authors commenting and supporting this change build core ecosystem software that is used by thousands of developers and dApps, like ENS, Optimism, WalletConnect, Stripe, Shopify, PancakeSwap, Sushi, LooksRare, and many more. If EIPs won't consider their feedback, who is this process for?

First off, I think it's worth pointing out that this process for those who choose to participate. Before this comment, @jxom had been the primary voice in opposition to removing the `providerId` property. It's good we've got more participation from yourself and @lochie because of this PR.

As for your question about concrete reasons why this change should not move forward I've highlighted one of my key concerns which is the centralization factor that a hierarchically namespaced identifier introduces. I'd also like to point out some of the more technical concerns.

1. This proposal assumes that a wallet won't change their identifier and doesn't specify why that can't be done. Under the current wording it can be reasonably interpreted that the wallet can change their identifier because the term "session" has not been well defined. Previously, the term "session" could reasonably be interpreted to mean for the lifetime of the page which aligns pretty well with how this term gets commonly used by browsers.
2. This proposal doesn't define the format of the string making it far more likely that an identifier namespace collision can occur or be accidentally mutated.
3. In terms of undefined behavior, it doesn't normatively define what DApp libraries should do if the identifier is changed, how to detect if multiple wallets are supplying the same identifier, nor how to handle if no identifier or an unknown identifier is supplied.
4. It doesn't state how wallets should respond in the case when one of these undefined behaviors occur either which means that we'll have a coordination problem on our hands when a bug occurs.

Ultimately this leaves a lot to be left for implementers to decide which is how this is currently handled with the `window.ethereum.is*` syntax that's currently relied upon today by DApp libraries. For the most part, this works too although it has lead to the case where when a new wallet appears they need to make sure they get supported by these core DApp libraries and the version needs to be updated within the DApp in order to show up. This effectively acts as a "registering" support for them today and is one of the primary reason Brave needed to create the [Wallet partner program](https://brave.com/october-wallet-partners/) so that the Brave icon would appear on DApps. Without that, users either thought that DApps didn't support our wallet. Alternatively, they needed to click Metamask in order to use our wallet which is part of the reason we supply the `isMetamask` property. The other being that the original implementation of Brave Wallet was just the Metamask extension built directly into Brave so for backwards compatibility we kept it even after we re-implemented Brave Wallet.

By leaving the providerId out, we can achieve a stateless design where any wallet that utilizes the Event loop properly can register to any DApp in an adhoc way without the need to register. Many of the additional UI features remove this stateless design and rely on the `providerId` property to managed additional state which I understand the value in. It can create a better UX for users. However, the requirement of the property creates a reliable assumption (albeit not very robust as defined in this proposal, nor my alternative proposal) of the existence of the property in such a way that DApps will begin to rely upon. This assumption of a stateful design based on the identifier means that we will have given up the decentralization that would be achievable with a stateless design. So, by leaving this property as optional my theory is that it will force DApp library developers and DApp implementers to not assume it's existence and therefore support a decentralized fallback so that even in the case of new wallets, they won't re-encounter the same issues that we ran into at Brave.

**Comment by tmm** [comment]
> I've put forth an alternative proposal which @jxom has approved so I'd prefer to get more concrete about the changes you're hoping to see added to the spec if you believe that what I've proposed is not sufficient.

Read through the proposal. Thank you for putting that together. It's unclear why reverse DNS was introduced. What is the benefit of that over another format? Domains come and go via expiration, seizure by authorities, switching to a different domain, etc. Further, since the `rdns` property is optional, it's not something library authors can rely on for the uses described by others in the previous comments. 

> My argument for this from the beginning has been that we should avoid a centralized registry by any means necessary.

Fine with me.

> With that said, I have no intention of ignoring the feedback you're providing and am noting that there's reservations from DApp library developers which is important to consider. 

Appreciate it!

> From the beginning I noticed that this spec was largely being driven by Wallet Devs so it's good that we're seeing participation from the DApp side (with the exception of Web3Modal who's been participating from the beginning) as well.

Makes sense that Wallet Devs are highly incentivized to push for a solution due to the pitfalls of `window.ethereum` and their wallets being unusable. Web3Modal is an abstraction on top of [wagmi](https://github.com/wagmi-dev/wagmi) (@jxom and @tmm are maintainers), where EIP-6963 will be implemented for developers to consume.

> As for your question about concrete reasons why this change should not move forward I've highlighted one of my key concerns which is the centralization factor that a hierarchically namespaced identifier introduces.

Again, we don't need to do this.

> I'd also like to point out some of the more technical concerns.
>
> 1. This proposal assumes that a wallet won't change their identifier and doesn't specify why that can't be done. Under the current wording it can be reasonably interpreted that the wallet can change their identifier because the term "session" has not been well defined. Previously, the term "session" could reasonably be interpreted to mean for the lifetime of the page which aligns pretty well with how this term gets commonly used by browsers.
>
> 2. This proposal doesn't define the format of the string making it far more likely that an identifier namespace collision can occur or be accidentally mutated.
>
> 3. In terms of undefined behavior, it doesn't normatively define what DApp libraries should do if the identifier is changed, how to detect if multiple wallets are supplying the same identifier, nor how to handle if no identifier or an unknown identifier is supplied.
>
> 4. It doesn't state how wallets should respond in the case when one of these undefined behaviors occur either which means that we'll have a coordination problem on our hands when a bug occurs.
>
> Ultimately this leaves a lot to be left for implementers to decide which is how this is currently handled with the window.ethereum.is* syntax that's currently relied upon today by DApp libraries. 

Agree that definitions and behaviors can be more detailed, but none of these technical concerns find fault with the concept of a `providerId`. I'm happy to work on improving those more (my TODO coming out of this).

The gist is picking a format and sticking with it (e.g. camelized/snakecase/other casing with no spaces, special characters/numbers allowed, etc. Brave Wallet could be `brave`, `braveWallet`, etc. Really just need to come up with a spec.)

> By leaving the providerId out, we can achieve a stateless design where any wallet that utilizes the Event loop properly can register to any DApp in an adhoc way without the need to register.

Not sure how including `providerId` makes this stateful. It's just another piece of info (chosen by wallets), like `icon` (required by 6963), that library authors/dApp developers downstream can use for things like server-side rendering hydration, targeting specific wallets, or for any other reason why having a stable identifier is valuable.

I'm sure it was also challenging to come up with a definition for chain IDs and there are still conflicts that pop up occasionally, but having a stable identifier for chains was likewise too important not to and overall good. We have the opportunity to do the same for wallets with `providerId` now.

**Comment by kdenhartog** [comment]
> Thank you for putting that together. It's unclear why reverse DNS was introduced. What is the benefit of that over another format?

To the best of my knowledge it was chosen as a way to maintain a globally unique identifier that mostly eliminates namespace collisions and doesn't require an explicit "registration" because the owner of the domain is assumed to control the identifier. @pedrouid was the person who introduced it originally so may be able to provide more context.

> Domains come and go via expiration, seizure by authorities, switching to a different domain, etc. 

Agreed, hence why I've argued that accounting for updating/deleting etc is necessary and is currently undefined behavior in both proposals.

> Again, we don't need to do this.

I think you'll find that if we don't explicitly do this, then library authors will implicitly do this through implementation which then means that what they choose to merge or not merge will act as the "governance" of the "psuedo-registry".

> Agree that definitions and behaviors can be more detailed, but none of these technical concerns find fault with the concept of a providerId. I'm happy to work on improving those more (my TODO coming out of this).
> 
> The gist is picking a format and sticking with it (e.g. camelized/snakecase/other casing with no spaces, special characters/numbers allowed, etc. Brave Wallet could be brave, braveWallet, etc. Really just need to come up with a spec.)

Identifier design inherently bakes in tradeoffs that impacts how they can be used and relied upon. I recommend taking a read of [Zooko's Triangle](https://en.wikipedia.org/wiki/Zooko's_triangle) to understand what I mean if you haven't already heard of that before.

As you can see from Zooko's trinagle it goes beyond just picking a format for the string to be parsed so let's break it down. Let's start with a single example to make this a bit more clear. If a wallet implementation either doesn't send the identifier for the wallet or sends a non-spec compliant identifier, what is the expected behavior of a DApp implementation? Here's 3 options I can think of what it can do:

1. It can throw an error to be handled elsewhere by the application logic of the DApp
2. It can hang the DApp (this is poor design - not saying it should be done, but it will accidentally be done if we don't specify how this should work)
3. It can fire a response back to the wallet to indicate the failure. How should that be done?
4. It could ignore the provider and opt not to display it

Options 1 and 4 are the most likely cases on how to handle this. Do we need to standardize this behavior so that all implementations pick one of the implementations? The answer is "probably not" here. That seems like an implementation specific detail. And if this is an implementation detail already why require it to be defined within the standard and bloat the spec?

> Not sure how including providerId makes this stateful. It's just another piece of info (chosen by wallets), like icon (required by 6963), that library authors/dApp developers downstream can use for things like server-side rendering hydration, targeting specific wallets, or for any other reason why having a stable identifier is valuable.

The icon has specifically been moved to be a data URI as well so that the string contains the contents completely and prevents any sort of concerns. If it's malformed than it's a GIGO scenario. The examples you provide all create a reliance on the stability of the identifier in a way that cannot be guaranteed across an untrusted (which is the case because the extension is an independent origin from the page itself) interface.

> It's just another piece of info (chosen by wallets)
I think this is a bit disingenuous. If it actually was just an arbitrary identifier that wasn't to be relied upon then we wouldn't need it. An identifier in the most common use cases exists in order to map to an arbitrary piece of information that's being stored hence making this stateful. 

> server-side rendering hydration
The images that the server is rendering - what happens if the wallets designs are updated? This is the state that's being mapped to by the identifier in this example hence my argument that this is a stateful design.

> targeting specific wallets
This introduces the gatekeeper effect that Brave encountered and I'm hoping new wallets will no longer have to encounter with the design choices we make within this spec. My hope is that we can design a spec so that libraries like WAGMI can implement this spec without the requirement of [wallet specific code paths](https://github.com/WalletConnect/web3modal/blob/a27dd8330db435dee5bfe8388d900c3f6f6393b2/packages/ui/src/utils/DataUtil.ts#L43-L50). I can see a path of how this is achieved, hence why I'm arguing to keep this optional. By keeping it optional within the spec we force implementations to consider the implications without needing to define how they should work. That means that libraries like web3modal could take option 4 from above, or they could take additional options. For example, web3modal could choose to load well known wallets that include an identifier first and render unknown wallets second. My concern here is that if we require it implementers will skip the step of thinking about "render unknown wallets". This leads to the libraries like web3modal only supporting a limited number of wallets and has the downstream effect that DApp's only support a limited number of wallets unless they keep updating their web3modal versions consistently. 

> I'm sure it was also challenging to come up with a definition for chain IDs and there are still conflicts that pop up occasionally, but having a stable identifier for chains was likewise too important not to and overall good. We have the opportunity to do the same for wallets with providerId now.

We also have the opportunity to introduce a lot of implicit problems and tradeoffs that I've highlighted which is what I'm trying to avoid. If we can propose a solution that can address these concerns then I'm all ears, but from my experience it's going to be a rabbit hole that's going to lead to maintainer burden that traverse from the wallets to the DApps and all the code in between. 

I think it's worth pointing out here that I understand the purpose in this and I don't disagree with it in principle. In fact, Brave intends to supply this provider identifier. With that in mind though, I'm arguing my position for the overall impact of the ecosystem and it's ability to remain decentralized, to keep the spec robust and minimally scoped, and because I believe that the best path forward for this is to make it an intentional extension point. 

In standards, these can be incredibly useful if you choose them wisely. The example, I pointed out previously (I think in telegram only though) is that OAuth2 was very intentional about not specifying the data model of tokens within it's spec. It could have very easily done so in a way to make it only work with JWTs but because it didn't this meant that SAML tokens could also be used, Opaque API keys could be used, Verifiable Credentials could be used, and who knows what else along the way can be used as well. My point here is that I don't believe specifying this is wise for us to do. I think it's a net negative on the spec and will lead to additional maintenance that can be avoided by leaving this as optional rather than requiring it so that DApp developers expect it to behave in a way that it may not and cause unintended side effects.

**Comment by glitch-txs** [comment]
I think what's stated here is actually to avoid using the wallet's name to add particular features to Dapp and libraries that help to track and identify each wallet to offer a nice UX, having this as optional feature would make it as unreliable (or even more maybe) than using the wallet's name itself.

I also agree with what's been said with reverse DNS.

Why don't we just specify the functionality of this ID and possible usages and let the wallets choose what type of ID to use? I think, as long as it's a string there won't be any issues on the dapp side.
